tests/v5/mqtt5_client_tests.c

```diff
@@ -6076,3 +6076,299 @@ static int s_mqtt5_client_outbound_alias_lru_success_a_b_c_br_cr_a_fn(struct aws
+        aws_array_list_get_at_ptr(&context->completion_callbacks, (void **)&callback, i);
+
+        if (callback->type != expected_callback->type) {
+            return false;
+        }
+
+        if (callback->error_code != expected_callback->error_code) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+static void s_add_completion_callback(
+    struct mqtt5_dynamic_operation_timeout_test_context *context,
+    enum aws_mqtt5_packet_type type,
+    int error_code) {
+    aws_mutex_lock(&context->fixture->lock);
+
+    struct mqtt5_operation_timeout_completion_callback callback_entry = {
+        .type = type, .error_code = error_code, .timepoint_ns = 0};
+
+    aws_high_res_clock_get_ticks(&callback_entry.timepoint_ns);
+
+    aws_array_list_push_back(&context->completion_callbacks, &callback_entry);
+
+    aws_mutex_unlock(&context->fixture->lock);
+
+    aws_condition_variable_notify_all(&context->fixture->signal);
+}
+
+static void s_timeout_test_publish_completion_fn(
+    enum aws_mqtt5_packet_type packet_type,
+    const void *packet,
+    int error_code,
+    void *complete_ctx) {
+    (void)packet_type;
+    (void)packet;
+
+    s_add_completion_callback(complete_ctx, AWS_MQTT5_PT_PUBLISH, error_code);
+}
+
+static void s_timeout_test_subscribe_completion_fn(
+    const struct aws_mqtt5_packet_suback_view *suback,
+    int error_code,
+    void *complete_ctx) {
+    (void)suback;
+
+    s_add_completion_callback(complete_ctx, AWS_MQTT5_PT_SUBSCRIBE, error_code);
+}
+
+static void s_timeout_test_unsubscribe_completion_fn(
+    const struct aws_mqtt5_packet_unsuback_view *unsuback,
+    int error_code,
+    void *complete_ctx) {
+    (void)unsuback;
+
+    s_add_completion_callback(complete_ctx, AWS_MQTT5_PT_UNSUBSCRIBE, error_code);
+}
+
+static bool s_all_timeout_operations_complete(void *arg) {
+    struct mqtt5_dynamic_operation_timeout_test_context *context = arg;
+
+    return aws_array_list_length(&context->completion_callbacks) == context->expected_callback_count;
+}
+
+static void s_wait_for_all_operation_timeouts(struct mqtt5_dynamic_operation_timeout_test_context *context) {
+    aws_mutex_lock(&context->fixture->lock);
+    aws_condition_variable_wait_pred(
+        &context->fixture->signal, &context->fixture->lock, s_all_timeout_operations_complete, context);
+    aws_mutex_unlock(&context->fixture->lock);
+}
+
+/*
+ * Tests a mixture of qos 0 publish, qos 1 publish, subscribe, and unsubscribe with override ack timeouts.
+ *
+ * qos 1 publish with 3 second timeout
+ * subscribe with 2 second timeout
+ * qos 0 publish
+ * unsubscribe with 4 second timeout
+ * qos 1 publish with 1 second timeout
+ *
+ * We expect to see callbacks in sequence:
+ *
+ *  qos 0 publish success
+ *  qos 1 publish failure by timeout after 1 second
+ *  subscribe failure by timeout after 2 seconds
+ *  qos 1 publish failure by timeout after 3 seconds
+ *  unsubscribe failure by timeout after 4 seconds
```

Comment:
Debatable: I feel that another operation with timeout set to 0 would be good for coverage, as we process 0 differently. But I don't see a way to add this easily. So, I just wanted to point it out, maybe you have an idea how it can be covered without complicating things.

Reply:
We could add a test where the client has a default and we submit with 0 and assert that it uses the default.